### ISSUE TITLE
False positive MixedAssignment (Unable to determine the type that $... is being assigned to)

### ISSUE THREAD

**Issue by mj4444ru** [vimeo/psalm#8035]
Everything is correct in the first case:

https://psalm.dev/r/86ec3569e6

In the second case, we get a false error:

https://psalm.dev/r/aabbc2cbbf

**Comment by kkmuffme** [comment]
Afaik this is correct. A non-empty-list can contain all types of variables, therefore this is a mixed assignment.
(therefore actually the first example should (could?) give MixedAssignment error too actually, but since you validate a type later I guess it doesn't)

**Comment by AndrolGenhald** [comment]
Yes, I think this might be sort of intentional. Some `mixed` errors are automatically suppressed if there is a type check immediately following the assignment, but I think right now only scope ending statements (`return`, `throw`, `exit`, etc) are considered for this analysis.

@orklah do you have any thoughts on allowing reassignment of a `mixed` assigned variable as a type check to suppress mixed issues? I'm thinking it could be nice to have, but it also seems a bit different from the way it works now so I'm not sure if it would be annoying and seem more like a false negative in certain instances.

**Comment by AndrolGenhald** [comment]
> any "is_string($var)" next to the mixed assignment suppress the issue

At first I thought that sounded like a terrible idea, but actually I think that might be a good way to go about it as long as it doesn't change type inference. That way [this](https://psalm.dev/r/daa711f44d) would remove the `MixedAssignment` but leave the `MixedReturnStatement`. That sounds like it might be difficult to implement as well though, but I haven't actually looked at how it works currently so maybe not.

**Comment by AndrolGenhald** [comment]
> I would also like the following code to suppress the error for the $s variable and determine its type inside the IF block

That part [already works](https://psalm.dev/r/8a42e994ba), it just doesn't suppress the `MixedAssignment` since `$s` might be used further down.

**Comment by weirdan** [comment]
> If I make comparisons, or pass to where mixed is also expected, this is normal and there is no point in this error. Mixed has the right to exist.

If you think so you just need to run Psalm at `errorLevel=2`. `errorLevel=1` (formerly called `totallyTyped=true`) only differs from `errorLevel=2` by not allowing mixed.

**Comment by mj4444ru** [comment]
It seems to me the logic of this error is very strange. Especially strange ways of suppressing it in the form of ```! is_*())```.
But it seems to be working as intended.

You can also suppress the error in this way:
```php
is_string(/** @var mixed */ $s = $arr[0])
```

Maybe add ```is_*())``` behavior similar to ```! is_*())```.

**Comment by mj4444ru** [comment]
> > If I make comparisons, or pass to where mixed is also expected, this is normal and there is no point in this error. Mixed has the right to exist.
> 
> If you think so you just need to run Psalm at `errorLevel=2`. `errorLevel=1` (formerly called `totallyTyped=true`) only differs from `errorLevel=2` by not allowing mixed.

In fact, I find most of the Level 1 mistakes very useful. The question is probably only about 1 error, and there are 23 of them.

**Comment by mj4444ru** [comment]
> @mj4444ru You can always use an [issue handler](https://psalm.dev/docs/running_psalm/dealing_with_code_issues/#config-suppression) to globally suppress all `MixedAssignment` issues if you don't think it's useful.

Thank you. I thought so..
